A few improvements to Operator overloading

karanchahal opened the pull request:
Now, as one can see in main.cpp we can run the sigmoid test with relatively few boilerplate code.

karanchahal committed:
Vector addition sample added

karanchahal committed:
Figured ot a way to solve for temporaries

karanchahal committed:
Built up functionality for more readable sigmoid test by overloading const tensors. Need to be aware of lhs and rhs while constructing an expression

uditarora commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
-build
\ No newline at end of file
+build
+a.out
```
Add newline at the end of the file?

Some diff tools have a weird way of handling files that don't end with newlines, so it's always better to add a newline at the end of code files.

uditarora replied:
Add newline at the end of the file?

Some diff tools have a weird way of handling files that don't end with newlines, so it's always better to add a newline at the end of code files.

uditarora commented on main.cpp:
```diff
@@ -18,7 +10,7 @@ void sigmoidTest() {
 
     Tensor<float> w3(vector<float>{-3});
-
+    
```
nit: Remove double newline?

uditarora commented on main.cpp:
```diff
@@ -44,8 +36,38 @@ void sigmoidTest() {
+//3. For the problems of temporaries
 int main() {
 
+ Tensor<float> w0(vector<float>{2});
```
nit: Incorrect indentation

uditarora commented on divideOperation.h:
```diff
@@ -7,7 +7,7 @@ template <typename T>
 
     public:
-        Tensor<T> *t1, *t2; // generally an operation had two operands
+        Tensor<T> *t1, *t2, *t3; // generally an operation had two operands
```
nit: move t3 declaration to next line like in addOperation.h for more clarity

uditarora replied:
nit: move t3 declaration to next line like in addOperation.h for more clarity

uditarora commented on exponentOperation.h:
```diff
@@ -8,7 +8,7 @@ template <typename T>
 
     public:
-        Tensor<T> *t1; // generally an operation had two operands
+        Tensor<T> *t1,*t3; // generally an operation had two operands
```
nit: move t3 to newline for better clarity, and add a comment indicating the input and temp/output tensors

uditarora commented on multiplyOperation.h:
```diff
@@ -6,7 +6,7 @@ template <typename T>
 
     public:
-        Tensor<T> *t1, *t2; // generally an operation had two operands
+        Tensor<T> *t1, *t2, *t3; // generally an operation had two operands
```
nit: move t3 to newline (as in other comments)

uditarora commented on operations_Impl.h:
```diff
@@ -14,7 +14,8 @@ void MultiplyOperation<T>::backward(vector<T> grad) {
 template <typename T>
 Tensor<T> MultiplyOperation<T>::forward() {
-    return Tensor<T>(this->t1->val*this->t2->val, this);
+    this->t3 = new Tensor<T>(this->t1->val*this->t2->val, this);
```
nit: should we add whitespace around 'multiply (*)' operator for more clarity?

uditarora commented on operations_Impl.h:
```diff
@@ -27,7 +28,8 @@ void AddOperation<T>::backward(vector<T> grad) {
 template <typename T>
 Tensor<T> AddOperation<T>::forward() {
-    return Tensor<T>(this->t1->val+this->t2->val, this);
+    this->t3 = new Tensor<T>(this->t1->val+this->t2->val, this);
```
nit: should we add whitespace around 'add (+)' operator for more clarity?

uditarora commented on operations_Impl.h:
```diff
@@ -44,7 +46,8 @@ void DivideOperation<T>::backward(vector<T> grad) {
 template <typename T>
 Tensor<T> DivideOperation<T>::forward() {
-    return Tensor<T>(this->t1->val*this->t2->val, this);
+    this->t3 = new Tensor<T>(this->t1->val*this->t2->val, this);
```
nit: should we add whitespace around 'multiply (*)' operator for more clarity?

uditarora commented on tensor.h:
```diff
@@ -28,25 +44,62 @@ class Tensor {
         this->backOp = op;
     }
 
+    bool isConstz(const Tensor<T> &one) {
```
Would it be better to make these isConstsz functions private?

nit: why have you added 'z' after isConst?

karanchahal replied:
Will remove these functions as they were just for debugging if "this" is a const Tensor or not.

uditarora commented on tensor.h:
```diff
@@ -19,6 +28,13 @@ class Tensor {
         // default
     }
 
+    Tensor(const Tensor<T> *two) {
```
nit: Add a comment explaining the purpose of this constructor?

uditarora commented on vecOps.cu:
```diff
@@ -0,0 +1,100 @@
+
+    return 1;
+    
+}
```
nit: add newline at end of file

uditarora commented on vecOps.cu:
```diff
@@ -0,0 +1,100 @@
+    cudaEventCreate(&launch_end);
+
+    // Warmup
+    add<<<n,1>>>(d_a, d_b, d_c);// num blocks, num_threads
```
Maybe we can add a boolean parameter to the function which controls whether we want to run it 100 times and get the average, or just run it once? It can perhaps have a default value set to 1 run.

karanchahal replied:
SO in a GPU function, which add is (signified by the <<< >>> operators), we can't signify how many times we want the kernel to run. We would have to probably wrap this in a separate function which controls the number of times this function is run. Should I do that ?

karanchahal replied:
And before testing timing on the GPU you have to do one warmup. How do you think this needs to be structured ?

uditarora replied:
I don't have enough context on GPU programming to answer this well.

My comment was more from a high-level perspective, that instead of running it 100 times for each function call, maybe just have a switch which controls the number of times the same operation is run (irrespective of warmup).

karanchahal replied:
Okay I'll do that man

karanchahal replied:
So I was testing the vector addition and some bugs were coming up that will take some time to fix. So I've deleted the vecOps.cu file containing the cuda stuff. I'll work on that separately in another branch. I've fixed everything else. Thanks for the great review!

karanchahal committed:
Cleanup of code, added comments, deleted redundant files and functions.

uditarora commented on exponentOperation.h:
```diff
@@ -8,7 +8,8 @@ template <typename T>
     public:
-        Tensor<T> *t1,*t3; // generally an operation had two operands
+        Tensor<T> *t1;
+        Tensor<T> *t3; // Input tensors
```
nit: Incorrectly placed comment

uditarora commented on main.cpp:
```diff
@@ -66,8 +61,14 @@ int main() {
 
-    sigmoidTest();
+    newSigmoidTest();
 }
```
nit: add newline

karanchahal commented:
Updated code with new commit having the fixes requested. The GPU code needs some more discussion on what has to be done. Also in the spirit of having a clean codebase and git history, after the changes are approved, should I smash all the above commits into 1 ? Or smash the commits that just clean up stuff and not add functionality into 1 ?

uditarora commented:
> Updated code with new commit having the fixes requested. The GPU code needs some more discussion on what has to be done. Also in the spirit of having a clean codebase and git history, after the changes are approved, should I smash all the above commits into 1 ? Or smash the commits that just clean up stuff and not add functionality into 1 ?

The second option sounds better to me. One commit for cleanup and other for new features/additions.

karanchahal commented:
Squashed the cleanup commits and deleted cuda file due to some errors. I'll address that later in some other issue. Maybe we can make an issue of that.
